lib/kafka_ex/new/kafka_ex_api.ex

```diff
@@ -0,0 +1,88 @@
+defmodule KafkaEx.New.KafkaExAPI do
+  @moduledoc """
+  This module interfaces Kafka through the ServerKayrock implementation
+
+  This is intended to become the future KafkaEx API
+  """
+
+  alias KafkaEx.ServerKayrock
+  alias KafkaEx.New.ClusterMetadata
+  alias KafkaEx.New.Topic
+
+  @type node_id :: non_neg_integer
+  @type topic_name :: binary
+  @type partition_id :: non_neg_integer
+  @type consumer_group_name :: binary
+  @type offset :: non_neg_integer
+  @type error_atom :: atom
+  @type client :: GenServer.server()
+  @type correlation_id :: non_neg_integer
+
+  @doc """
+  Fetch the latest offset for a given partition
+  """
+  @spec latest_offset(client, topic_name, partition_id) ::
+          {:error, error_atom} | {:ok, offset}
+  def latest_offset(client, topic, partition) do
+    request = %Kayrock.ListOffsets.V1.Request{
+      replica_id: -1,
+      topics: [
+        %{topic: topic, partitions: [%{partition: partition, timestamp: -1}]}
+      ]
+    }
+
+    {:ok, resp} =
+      ServerKayrock.kayrock_call(
+        client,
+        request,
+        {:topic_partition, topic, partition}
+      )
+
+    [topic_resp] = resp.responses
+    [%{error_code: error_code, offset: offset}] = topic_resp.partition_responses
+
+    case error_code do
+      0 -> {:ok, offset}
+      _ -> {:error, Kayrock.ErrorCode.code_to_atom(error_code)}
+    end
+  end
+
+  @doc """
+  Get topic metadata for the given topics
+
+  Fetches metadata from the server if necessary.  Set allow_topic_creation to
```

Comment:
Do we allow auto-creating topics now if we request the metadata from them?

It also seems like this always fetches metadata from the server?

Reply:
Yes, we auto-create by default in the existing implementation.  Before metadata request v4 that was the default.

This does always fetch from the brokers.  It seems like the only use cases for this call are either tests or ops-type things, so it doesn't make sense to me to just return the cached metadata.  Especially since we may or may not have the requested topic(s) cached.  (If you didn't catch it, I disabled fetching metadata for _every_ topic - I was actually finding that to cause timeouts with our cluster at work because it has so many topics.  The kayrock client implementation fetches metadata on-demand for topics.)

Reply:
ok, makes sense. I guess just a tweak to the comment so that it's more clear that it's always doing a fetch from the brokers for the metadata for the topic(s) and I'm 👍